README.md
<!-- Copyright 2015. Author: Jeffrey Hing. All Rights Reserved. MIT License -->

# AngularWidthQuery

AngularWidthQuery is a set of directives which applies HTML classes to 
elements based upon a container element's width. 

It was created to support the fluid layout of a container's elements as its
width changed due to the opening and closing of other panels in the UI.

It replaces the need for CSS media queries in situations where 
the container element's width determines the layout, rather than the screen width.

## Table of Contents

- [Features](#features)
- [Installation](#installation)
- [Example](#example)
- [Usage](#usage)
    - [Factory Method](#factory-method)
    - [WidthQuery Directive](#widthquery-directive)
    - [WidthQueryClass Directive](#widthqueryclass-directive)
   
## Features

* Less than 4K minified.
* CSS transition and animation aware.
* Customizable directive names.
* Compatible with CommonJS, AMD, and non-module build environments.

## Installation

To install the package:

    npm install angular-width-query
    
To require the package:    

```javascript
var widthQueryDirective = require("angular-width-query");
```     

## Example

The easiest way to understand AngularWidthQuery is to walk 
through a quick example of creating a fluid layout consisting of 6 
equally sized sections. At small width, the sections are stacked. 
At medium width, the sections are arranged in two columns. At large width,
the sections are arranged in three columns.

### 1. Add the Directives

First, add the directives from AngularWidthQuery to an Angular module.

```javascript
widthQueryDirective('app');
```    

### 2. Create the HTML Structure

Next, create the HTML structure for the container and the 6 sections.

```html
<div width-query="container">
    <div width-query-class="section">Section 1</div>
    <div width-query-class="section">Section 2</div>
    <div width-query-class="section">Section 3</div>
    <div width-query-class="section">Section 4</div>
    <div width-query-class="section">Section 5</div>
    <div width-query-class="section">Section 6</div>
</div>
```

The values passed to the directives are the base HTML classes that will
be initially added to the elements.

Based upon the current container element width, classes are dynamically 
added to the elements to reflect the current width. For example, 
"container-small" and "section-small" will be added to their respective 
elements when the width of the container element is considered to be small.

The width ranges and class modifiers are fully customizable. See the Usage
section below for details.

### 3. Create the CSS

Lastly, create the CSS rules.

```css
.container {
    box-sizing: border-box;
    font-size: 0;
}

.section {
    border: 1px solid #c0c0c0;
    box-sizing: border-box;
    display: inline-block;
    font-size: 1rem;
    height: 300px;
}

.section-large {
    width: 33.33%;
}

.section-medium {
    width: 50%;
}

.section-small {
    width: 100%;
}
```

That's it. You're done. 

## Usage

### Factory Method

The `widthQueryDirective()` factory method adds the AngularWidthQuery
directives to an Angular module.

To add the directives, call `widthQueryDirective()` with
the name of the Angular module, and any options:

```javascript
widthQueryDirective('app', {
    directiveNames: {
        widthQuery: <string>,
        widthQueryClass: <string>
    },
    modifiers: [
        [<string>, <number>, <number>],
        ...
    ],
    pollingInterval: <milliseconds>
});
```
#### 'directiveNames' option

This option allows the names of the widthQuery and widthQueryClass directives
to be changed to suit your needs.

```javascript
widthQueryDirective('app', {
    directiveNames: {
        widthQuery: 'myWidthQuery',
        widthQueryClass: 'myWidthQueryClass'
    }
});
```

#### 'modifiers' option

This option allows you to specify the class modifiers that can be added to the
base HTML class.

Each item in the array is an array consisting of the class modifier, 
the lower width range, and the upper width range.

```javascript
widthQueryDirective('app', {
    modifiers: [
        ['small', 0, 767],
        ['medium', 768, 1023],
        ['large', 1024, 100000]
    ]
});
```

#### 'pollingInterval' option

This option allows you to specify the interval at which the directive polls
the element's width. The directive begins polling the element's width when
a window resize event or Angular digest event occurs. When the width stops 
changing, the polling stops. 

```javascript
widthQueryDirective('app', {
    pollingInterval: 100
});
```
If you have no animations or transitions, you can set this to 0.

### WidthQuery Directive

The widthQuery directive updates the element's HTML class based upon the 
element's width.

```html
<div width-query="container">
    ...
</div>
```

The widthQuery directive can be passed the base HTML class, or a set of
options. These options override the options passed to the factory method.

```html
<div width-query="{
    class: 'container',
    modifiers: [
        ['small', 0, 399],
        ['medium', 400, 1279],
        ['large', 1280, 100000]
    ],
    pollingInterval: 100,
    widthPollingListener: 'ctrl.widthChanged(width)',
    widthListener: 'ctrl.widthChanged(width)'
}">...</div>
```

#### 'class' option

This option specifies the base HTML class to initially add to the element.

```html
<div width-query="{
    class: 'container'
}">...</div>
```

#### 'modifiers' option

This option allows you to specify the class modifiers that can be added to the
base HTML class.

Each item in the array is an array consisting of the class modifier, 
the lower width range, and the upper width range.

```html
<div width-query="{
    modifiers: [
        ['small', 0, 767],
        ['medium', 768, 1023],
        ['large', 1024, 100000]
    ]
}">...</div>
```

#### 'pollingInterval' option

This option allows you to specify the interval at which the directive polls
the element's width. The directive begins polling the element's width when
a window resize or Angular digest event occurs. When the width stops 
changing, the polling stops. 

```html
<div width-query="{
    pollingInterval: 100
}">...</div>
```
If you have no animations or transitions, you can set this to 0.

#### 'widthListener' option

This option allows you to specify a function to call when the width has
stopped changing. This is useful when you need to update the display of other
elements when an animation or transition has completed.

```html
<div width-query="{
    widthListener: 'ctrl.widthChanged(width)'
}">...</div>
```

#### 'widthPollingListener' option

This option allows you to specify a function to call while the width is 
changing. This is useful when you need to update the display of other
elements during an animation or transition.

```html
<div width-query="{
    widthPollingListener: 'ctrl.widthChanged(width)'
}">...</div>
```

### WidthQueryClass Directive

The widthQueryClass directive updates the element's HTML class based upon the 
state of the parent widthQuery directive.

In the example below, if the widthQuery directive has applied class 
"container-small" to its element, the widthQueryClass directive will have 
applied class "section-small" to its element.

```html
<div width-query="container">
    <div width-query-class="section">...</div>
    ...
</div>
```

The widthQuery directive is passed the base HTML class to initially add to 
the element.


CHANGELOG.md
### 0.1.3 (2016-10-04)
* Fix README documentation.

### 0.1.2 (2016-09-20)
* Package description and README update.

### 0.1.1 (2016-09-17)
* README update.

### 0.1.0 (2016-09-16)
* Initial version.

src/utilities/test/HasInput.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License
 *
 * Provides convenience functions to emulate mouse input for a jquery
 * wrapped element.
 */
'use strict';

//-------------------------------------
// Module exports
//-------------------------------------

module.exports = HasMouseInput;

//-------------------------------------
// Module dependencies and variables
//-------------------------------------

// Private model name.
var MODEL = '_hasMouseInput';

//-------------------------------------
// HasMouseInput class
//-------------------------------------

/*
 * @constructor
 */
function HasMouseInput(element) {
    var m = this[MODEL] = {};

    m.element = element;
}

var proto = HasMouseInput.prototype;

proto.mouseClick = function() {
    var m = this[MODEL];
    m.element.trigger('click');
};

proto.mouseEnter = function() {
    var m = this[MODEL];
    m.element.trigger('mouseenter');
};

proto.mouseLeave = function() {
    var m = this[MODEL];
    m.element.trigger('mouseleave');
};

proto.focus = function() {
    var m = this[MODEL];
    m.element.triggerHandler('focus');
};

proto.keyPress = function(char) {
    var m = this[MODEL];
    m.element.trigger({
        type: 'keypress',
        which: char.charCodeAt(0)
    });
};

proto.keyPressEnter = function() {
    var m = this[MODEL];
    m.element.trigger({
        type: 'keypress',
        which: 13
    });
};


src/utilities/test/HasStyle.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License
 *
 * Provide convenience functions to manaage the elements style.
 */
'use strict';

//-------------------------------------
// Module exports
//-------------------------------------

module.exports = HasStyle;

//-------------------------------------
// Module dependencies and variables
//-------------------------------------

// Private model name.
var MODEL = '_hasStyle';

//-------------------------------------
// HasStyle class
//-------------------------------------

/*
 * @constructor
 *
 * @param {object} element
 */
function HasStyle(element) {
    var m = this[MODEL] = {};

    m.element = element;
}

var proto = HasStyle.prototype;

/*
 * Indicates whether the element has the specified element class.
 *
 * @param {string} elementClass
 * @return {boolean}
 */
proto.hasClass = function(elementClass) {
    var m = this[MODEL];
    return m.element.hasClass(elementClass);
};


src/utilities/test/allTests.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License
 *
 * Load all tests ending in '.test.js' in a single bundle.
 */
'use strict';

(function() {
    var testsContext = require.context('../..', true, /\.test\.js$/);
    testsContext.keys().forEach(testsContext);

    /* Print out test files
    testsContext.keys().forEach(function(key) {
        console.log(key);
    });
    */
})();


run.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License
 *
 * Build commmands.
 */
'use strict';

//-------------------------------------
// Dependencies and variables
//-------------------------------------

var CommandQueue = require('command-queue');

// Clean distribution command.
var cleanDistCmd = new CommandQueue().sync('rimraf dist', 'mkdir dist');

//
// Test commands
//

// Test source file.
var testCmd = 'karma start --single-run --#kdev';

// Test source file and watch for changes.
var testWatchCmd = 'karma start --#kdev';

// Test distribution file.
var testDistCmd = 'karma start --single-run --#kdist';

// Test minimized distribution file.
var testDistMinCmd = 'karma start --single-run --#kdistMin';

//
// Compile commands
//

// Create distribution file.
var webpackDistCmd = 'webpack --progress --#wdist';

// Create minified distribution file.
var webpackDistMinCmd = 'webpack -p --progress --#wdistMin';

// Run webpack dev server and watch for changes.
var webpackWatchCmd = 'webpack-dev-server --hot --#wdev';

//-------------------------------------
// Run build commands
//-------------------------------------

process.env.PATH += ';node_modules/.bin';

var queue = new CommandQueue();

switch (process.argv[2]) {
    case 'dev':
        queue.async(webpackWatchCmd, testWatchCmd).run();
        break;
    case 'test':
        queue.sync(testCmd).run();
        break;
    case 'dist':
        queue.sync(
            cleanDistCmd, testCmd,
            webpackDistCmd, testDistCmd,
            webpackDistMinCmd, testDistMinCmd).run();
        break;
}


src/app.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License
 *
 * Application entry point.
 */
'use strict';

//-------------------------------------
// Initialize Angular
//-------------------------------------

require('angular');
require('angular-ui-router');

angular.module('app', ['ui.router']);

//-------------------------------------
// Initialize core UI
//-------------------------------------

require('./app.css');
require('angular-width-query')('app');
require('./views/example/view.css');

//-------------------------------------
// Initialize routes
//-------------------------------------

angular.module('app').config(function($stateProvider, $urlRouterProvider) {

    $urlRouterProvider.otherwise('/');

    $stateProvider
        .state('example', {
            url: '/',
            controller: require('./views/example/viewController.js'),
            template: require('./views/example/view.html'),
            controllerAs: 'ctrl'
        });
});


karma.conf.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License
 *
 * Karma configurations.
 */
'use strict';

//-------------------------------------
// Dependencies and variables
//-------------------------------------

var flags = require('minimist')(process.argv.slice(2));
var path = require('path');
var karmaWebpackPlugin = require('karma-webpack');
var webpackConfig = require('./webpack.config');

// Additional functions needed for testing.
var TEST_UTILITIES_DIR = 'src/utilities/test';

/*
 * Creates a karma configuration.
 *
 * @param {string} sourceFile The source file to test
 * @param {array} loaders An array of loaders to apply to the source file.
 */
function createConf(sourceFile, loaders) {

    var allLoaders = [];
    var testFilesPattern = path.join(TEST_UTILITIES_DIR, 'allTests.js');

    if (loaders) {
        allLoaders.concat(loaders);
    }

    // Return a new instance each time.
    var conf = {

        browsers: ['PhantomJS'],

        files: [
            // https://github.com/webpack/style-loader/issues/31
            'node_modules/phantomjs-polyfill/bind-polyfill.js',
            'node_modules/jquery/dist/jquery.js',
            'node_modules/angular/angular.js',

            testFilesPattern
        ],

        frameworks: ['jasmine'],

        plugins: [
            karmaWebpackPlugin,
            'karma-jasmine',
            'karma-phantomjs-launcher'
        ],

        preprocessors: {},

        reporters: [
            'dots'
        ],

        webpack: {
            module: {
                loaders: allLoaders
            },
            resolve: {
                alias: {},
                fallback: [
                    TEST_UTILITIES_DIR
                ]
            }
        },

        webpackMiddleware: {
            noInfo: true
        }

    };

    // Resolve to the correct file for testing.
    conf.webpack.resolve.alias[webpackConfig.library.projectName] =
        path.join(__dirname, sourceFile);

    conf.preprocessors[testFilesPattern] = ['webpack'];

    return conf;
}

//-------------------------------------
// Exports
//-------------------------------------
if (flags['#kdist']) {
    // Test distribution file.
    module.exports = function(config) {
        config.set(createConf(
            'dist/' + webpackConfig.library.filename
        ));
    };

} else if (flags['#kdistMin']) {
    // Test minimized distribution file.
    module.exports = function(config) {
        config.set(createConf(
            'dist/' + webpackConfig.library.filenameMin
        ));
    };

} else if (flags['#kdev']) {
    // Test source file.
    module.exports = function(config) {
        config.set(createConf(
            webpackConfig.library.sourceFile,
            webpackConfig.library.sourceLoaders
        ));
    };
}


webpack.config.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License
 *
 * Webpack configurations.
 */
'use strict';

//-------------------------------------
// Dependencies and variables
//-------------------------------------

var flags = require('minimist')(process.argv.slice(2));
var ip = require('ip');
var path = require('path');

// Set to true to expose development server.
var isServerPublic = false;

// Webpack loaders.
var loaders = {

    // Load CSS as javascript.
    css: {
        test: /\.css$/,
        loader: 'style!css!csslint?failOnError=false'
    },

    // Lint javascript.
    eslint: {
        test: /\.js$/,
        loader: 'eslint-loader',
        exclude: /node_modules/
    },

    // Load HTML as javascript.
    html: {
        test: /\.html$/,
        loader: 'html'
    }
};

// Library settings.
var library = {

    // GitHub project name.
    projectName: 'angular-width-query',

    // Name of library variable for non-module build environments.
    variable: 'widthQueryDirective',

    // Filename of library.
    filename: 'widthQueryDirective.js',

    // Filename of minimized library.
    filenameMin: 'widthQueryDirective.min.js',

    // Path to library source.
    sourceFile: './src/directives/widthQuery/directive.js',

    // Loaders to load the library source.
    sourceLoaders: [
        loaders.eslint
    ]
};

/*
 * Creates a webpack development configuration.
 */
function createDevConfig() {

    var config = {

        devServer: {
            host: isServerPublic ? ip.address() : undefined,
            contentBase: 'src/',
            noInfo: true,
            inline: true
        },

        entry: './src/app.js',

        eslint: {
            failOnError: false
        },

        module: {
            loaders: [
                loaders.css,
                loaders.eslint,
                loaders.html
            ]
        },

        output: {},

        resolve: {
            alias: {}
        }
    };

    config.resolve.alias[library.projectName] =
        path.join(__dirname, library.sourceFile);

    return config;
}

/*
 * Creates a webpack distribution configuration.
 *
 * @param {string} libraryName
 */
function createDistConfig(libraryName) {

    return {
        entry: library.sourceFile,

        eslint: {
            failOnError: true
        },

        module: {
            loaders: library.sourceLoaders
        },

        output: {
            filename: libraryName,
            library: library.variable,
            libraryTarget: 'umd',
            path: 'dist/'
        }
    };
}

//-------------------------------------
// Exports
//-------------------------------------

if (flags['#wdist']) {
    // Create distribution file.
    module.exports = createDistConfig(library.filename);

} else if (flags['#wdistMin']) {
    // Create minimized distribution file.
    module.exports = createDistConfig(library.filenameMin);

} else if (flags['#wdev']) {
    // Create development bundle.
    module.exports = createDevConfig();

} else {
    // Common settings.
    module.exports = {
        library: library,
        loaders: loaders
    };
}


src/directives/widthQuery/pageObject.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License
 *
 * Page object for width-query directive.
 */
'use strict';

//-------------------------------------
// Dependencies and variables
//-------------------------------------

var SNAKE_CASE_REGEXP = /[A-Z]/g;

var AngularTestContext = require('angular-test-context');
var directive = require('./directive');

// Private model name.
var MODEL = '_widthQueryPageObject';

//-------------------------------------
// Exports
//-------------------------------------

module.exports = WidthQueryPageObject;

//-------------------------------------
// Page object
//-------------------------------------

/*
 * @constructor
 */
function WidthQueryPageObject(html, scope, config) {

    var m = this[MODEL] = {};

    m.directiveName = 'width-query';
    m.childDirectiveName = 'width-query-class';

    if (config && config.directiveNames) {
        m.directiveName =
            this.snakeCase(config.directiveNames.widthQuery);
        m.childDirectiveName =
            this.snakeCase(config.directiveNames.widthQueryClass);
    }

    // Add directive to module.
    angular.module('testApp', []);
    directive('testApp', config);

    // Create test context.
    m.testContext = new AngularTestContext('testApp');

    // Compile the HTML snippet.
    m.element = m.testContext.compile(html, scope);
    $(document.body).append(m.element);

    m.testContext.digest();

    // Huh ?!!!. By accessing this variable, the CSS transitions
    // are now working in phantomJS.
    m.element[0].offsetWidth; // eslint-disable-line no-unused-expressions

    m.childElement = m.element.find('[' + m.childDirectiveName + ']');
}

var proto = WidthQueryPageObject.prototype;

proto.hasClass = function(className) {
    var m = this[MODEL];
    return m.element.hasClass(className);
};

proto.hasChildClass = function(className) {
    var m = this[MODEL];
    return m.childElement.hasClass(className);
};

proto.setWidth = function(width) {
    var m = this[MODEL];
    m.element.css('width', width + 'px');
    m.testContext.digest();
};

proto.cleanup = function() {
    var m = this[MODEL];
    $('[' + m.directiveName + ']').remove();
};

/*
 * Converts name from camel case to snake case.
 *
 * Modified angular code snippet.
 * @param {string} name
 * @return {string}
 */
proto.snakeCase = function(name) {
    return name.replace(SNAKE_CASE_REGEXP, function(letter, pos) {
        return (pos ? '-' : '') + letter.toLowerCase();
    });
};


src/directives/widthQuery/directive.js
/*
 * Copyright 2015. Author: Jeffrey Hing. All Rights Reserved.
 *
 * MIT License.
 *
 * The widthQuery directive updates the element's HTML class based upon
 * the element's width.
 *
 * @example
 *
 * widthQueryDirective('app', {
 *    directiveNames: {
 *        widthQuery: 'widthQuery',
 *        widthQueryClass: 'widthQueryClass'
 *    },
 *    modifiers: [
 *        ['small', 0, 767],
 *        ['medium', 768, 1023],
 *        ['large', 1024, 100000]
 *    ],
 *    pollingInterval: 100
 * });
 *
 * @example
 *
 * <div width-query="container">
 *     <div width-query-class="section">...<div>
 *     <div width-query-class="section">...<div>
 * </div>
 *
 * @example
 *
 * <div width-query="{
 *    class: 'container',
 *    modifiers: [
 *        ['small', 0, 399],
 *        ['medium', 400, 1279],
 *        ['large', 1280, 100000]
 *    ],
 *    pollingInterval: 100,
 *    widthListener: 'ctrl.widthChanged(width)',
 *    widthPollingListener: 'ctrl.widthChanging(width)'
 * }">
 *
 */
'use strict';

//-------------------------------------
// Dependencies and variables
//-------------------------------------

var widthQueryClassDirective = require('./widthQueryClass/directive');

// Private model name.
var MODEL = '_widthQueryDirective';

//-------------------------------------
// Exports
//-------------------------------------

module.exports = function(moduleName, options) {

    var config = {
        'class': null,
        directiveNames: {
            widthQuery: 'widthQuery',
            widthQueryClass: 'widthQueryClass'
        },
        modifiers: [
            ['small', 0, 767],
            ['medium', 768, 1023],
            ['large', 1024, 100000]
        ],
        pollingInterval: 50,
        maxPollingRetrys: 2,
        widthPollingListener: null,
        widthListener: null
    };

    if (options) {
        angular.merge(config, options);
    }

    var directiveNames = config.directiveNames;

    angular.module(moduleName).directive(directiveNames.widthQuery,
        function() {
            return widthQueryDirective(config);
        });

    widthQueryClassDirective(moduleName, config);
};

//-------------------------------------
// Directive controller
//-------------------------------------

Controller.$inject = ['$scope', '$element', '$attrs', '$window', '$timeout'];

/*
 * @constructor
 */
function Controller($scope, $element, $attrs, $window, $timeout,
    config) {

    var m = this[MODEL] = new ControllerModel($scope, $element,
        $timeout, config);

    // Parse attribute values.
    var values = $attrs[config.directiveNames.widthQuery];
    if (values) {
        if (values[0] === '[') {
            var arr = $scope.$eval(values);
            m.config.class = arr[0];
            m.config.widthListener = arr[1];
        } else if (values[0] === '{') {
            angular.extend(m.config, $scope.$eval(values));
        } else {
            m.config.class = values;
        }
    }

    // Watch for window resizing.
    var window = angular.element($window);
    window.bind('resize', startPolling);

    // Watch for user event.
    $scope.$watch(function() {
        startPolling();
    }, function() {});

    // Clean Up.
    $scope.$on('$destroy', function() {
        m.cancelPendingNotification();
        window.unbind('resize', startPolling);
    });

    function startPolling() {
        m.startPolling();
    }
}

var proto = Controller.prototype;

/*
 * Add a listener to notify of class modifier changes.
 *
 * @return {String} The initial class modifier to use.
 */
proto.addModifierListener = function(listener) {
    var m = this[MODEL];
    m.classModifierListeners.push(listener);
    return m.initialModifier;
};

/*
 * Set the initial class modifier and listen for class modifier changes.
 */
proto.setInitialModifier = function(element) {
    var m = this[MODEL];
    var modifiedClass;
    var className = '';

    // Add class to element.
    if (m.config.class) {
        element.addClass(m.config.class);
        className = m.config.class + '-';
    }

    m.initialModifier = m.calculateModifier(m.element.offsetWidth);

    // Listen for class modifier changes.
    addModifiedClass(this.addModifierListener(addModifiedClass));

    function addModifiedClass(modifier) {
        if (modifiedClass) {
            element.removeClass(modifiedClass);
        }
        modifiedClass = className + modifier;
        element.addClass(modifiedClass);
    }
};

//-------------------------------------
// Directive controller model
//-------------------------------------

/*
 * @constructor
 */
function ControllerModel(scope, element, timeout, config) {
    this.scope = scope;
    this.element = element[0];
    this.timeout = timeout;
    this.classModifierListeners = [];
    this.pendingNotification = null;
    this.width = null;
    this.config = config;
    this.pollingRetry = 0;
}

var modelProto = ControllerModel.prototype;

/*
 * Cancel any pending class notification.
 */
modelProto.cancelPendingNotification = function() {
    if (this.pendingNotification) {
        this.timeout.cancel(this.pendingNotification);
    }
};

/*
 * When the width stops changing, apply the new class name.
 */
modelProto.startPolling = function() {
    var self = this;

    if (self.pendingNotification) {
        return;
    }

    // Update class modifiers while width is changing
    // rather than waiting till the end to give a more
    // fluid appearance.
    var width = self.element.offsetWidth;
    if (width !== self.width) {
        self.notifyModifierListeners(width);
        self.notifyWidthPollingListener(width);
        self.pollingRetry = 0;
        self.width = width;
    }

    self.pendingNotification = self.timeout(function() {
        var newWidth = self.element.offsetWidth;

        self.pendingNotification = null;

        // Width is still changing
        if (newWidth !== self.width) {
            self.startPolling();
        } else {
            if (self.pollingRetry === 0) {
                self.notifyWidthListener(newWidth);
            }

            // Try a couple more times to handle race conditions.
            self.pollingRetry++;
            if (self.pollingRetry <= self.config.maxPollingRetrys) {
                self.startPolling();
            } else {
                self.pollingRetry = 0;
            }
        }

    }, self.config.pollingInterval, false);
};

/*
 * Return the class modifer based upon the width.
 *
 * @return {string} the class modifier
 */
modelProto.calculateModifier = function(width) {
    var modifiers = this.config.modifiers;
    var length = modifiers.length;
    for (var i = 0; i < length; i++) {
        var modifier = modifiers[i];
        if (width >= modifier[1] && width <= modifier[2]) {
            return modifier[0];
        }
    }
    return '';
};

/*
 * Notify listeners to use the calculated class modifier.
 */
modelProto.notifyModifierListeners = function(width) {
    var classModifier = this.calculateModifier(width);

    var listeners = this.classModifierListeners;
    var length = listeners.length;
    for (var i = 0; i < length; i++) {
        listeners[i](classModifier);
    }
};

/*
 * Notify listener that the width has changed.
 */
modelProto.notifyWidthListener = function(width) {
    if (this.config.widthListener) {
        this.scope.$eval(this.config.widthListener, {
            width: width
        });
    }
};

/*
 * Notify listener that the width has changed during the polling period.
 */
modelProto.notifyWidthPollingListener = function(width) {
    if (this.config.widthPollingListener) {
        this.scope.$eval(this.config.widthPollingListener, {
            width: width
        });
    }
};

//-------------------------------------
// Directive
//-------------------------------------

function widthQueryDirective(config) {
    ProxyController.$inject = ['$scope', '$element', '$attrs',
        '$window', '$timeout'
    ];

    function ProxyController($scope, $element, $attrs, $window, $timeout) {
        Controller.call(this, $scope, $element,
            $attrs, $window, $timeout, angular.merge({}, config));
    }
    ProxyController.prototype = Controller.prototype;

    return {
        controller: ProxyController,
        link: {
            pre: function(scope, element, attrs, controller) {
                controller.setInitialModifier(element);
            }
        },
        require: config.directiveNames.widthQuery
    };

}
